File: usecases/src/main/java/com/zeyad/usecases/services/GenericJobServiceLogic.java
package com.zeyad.usecases.services;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.util.Log;

import com.zeyad.usecases.Config;
import com.zeyad.usecases.services.jobs.FileIO;
import com.zeyad.usecases.services.jobs.Post;
import com.zeyad.usecases.stores.CloudDataStore;
import com.zeyad.usecases.utils.Utils;

import io.reactivex.Completable;

class GenericJobServiceLogic {
    GenericJobServiceLogic() {
    }

    Completable startJob(@NonNull Bundle extras, CloudDataStore cloudDataStore, Utils utils, String log) {
        if (extras.containsKey(GenericJobService.PAYLOAD)) {
            int trailCount = extras.getInt(GenericJobService.TRIAL_COUNT);
            switch (extras.getString(GenericJobService.JOB_TYPE, "")) {
                case GenericJobService.POST:
                    Log.d(GenericJobServiceLogic.class.getSimpleName(),
                            String.format(log, GenericJobService.POST));
                    return new Post(null, extras.getParcelable(GenericJobService.PAYLOAD),
                            Config.getApiConnection(), trailCount, utils)
                            .execute();
                case GenericJobService.DOWNLOAD_FILE:
                    Log.d(GenericJobServiceLogic.class.getSimpleName(),
                            String.format(log, GenericJobService.DOWNLOAD_FILE));
                    return new FileIO(trailCount, extras.getParcelable(GenericJobService.PAYLOAD),
                            null, true, cloudDataStore, utils)
                            .execute();
                case GenericJobService.UPLOAD_FILE:
                    Log.d(GenericJobServiceLogic.class.getSimpleName(),
                            String.format(log, GenericJobService.UPLOAD_FILE));
                    return new FileIO(trailCount, extras.getParcelable(GenericJobService.PAYLOAD),
                            null, false, cloudDataStore, utils)
                            .execute();
                default:
                    return Completable.complete();
            }
        }
        return Completable.complete();
    }
}

File: usecases/src/main/java/com/zeyad/usecases/services/GenericJobServiceLogic.kt
package com.zeyad.usecases.services

import android.os.Build
import android.os.Bundle
import android.support.annotation.RequiresApi
import android.util.Log
import com.zeyad.usecases.Config
import com.zeyad.usecases.services.jobs.FileIO
import com.zeyad.usecases.services.jobs.Post
import com.zeyad.usecases.stores.CloudStore
import io.reactivex.Completable

class GenericJobServiceLogic {

    @RequiresApi(Build.VERSION_CODES.N)
    fun startJob(extras: Bundle, cloudStore: CloudStore, log: String): Completable {
        if (extras.containsKey(GenericJobService.PAYLOAD)) {
            val trailCount = extras.getInt(GenericJobService.TRIAL_COUNT)
            when (extras.getString(GenericJobService.JOB_TYPE, "")) {
                GenericJobService.POST -> {
                    Log.d(GenericJobServiceLogic::class.java.simpleName,
                            String.format(log, GenericJobService.POST))
                    return Post(Config.context, extras.getParcelable(GenericJobService.PAYLOAD),
                            Config.apiConnection!!, trailCount)
                            .execute()
                }
                GenericJobService.DOWNLOAD_FILE -> {
                    Log.d(GenericJobServiceLogic::class.java.simpleName,
                            String.format(log, GenericJobService.DOWNLOAD_FILE))
                    return FileIO(trailCount, extras.getParcelable(GenericJobService.PAYLOAD),
                            Config.context, true, cloudStore)
                            .execute()
                }
                GenericJobService.UPLOAD_FILE -> {
                    Log.d(GenericJobServiceLogic::class.java.simpleName,
                            String.format(log, GenericJobService.UPLOAD_FILE))
                    return FileIO(trailCount, extras.getParcelable(GenericJobService.PAYLOAD),
                            Config.context, false, cloudStore)
                            .execute()
                }
                else -> return Completable.complete()
            }
        }
        return Completable.complete()
    }
}

